UI/MapWnd.cpp

```diff
@@ -5824,70 +5836,60 @@ void MapWnd::Sanitize() {
-    bool some_subscreen_was_visible = false;
-
-    if (m_sitrep_panel->Visible()) {
-        ToggleSitRep();
-        some_subscreen_was_visible = true;
-    }
-
-    if (m_research_wnd->Visible()) {
-        ToggleResearch();
-        some_subscreen_was_visible = true;
-    }
+void MapWnd::PushWndStack(std::shared_ptr<GG::Wnd> wnd) {
+    RemoveFromWndStack(wnd);
+    m_wnd_stack.push_back(wnd);
+}
 
-    if (m_design_wnd->Visible()) {
-        ToggleDesign();
-        some_subscreen_was_visible = true;
+void MapWnd::RemoveFromWndStack(std::shared_ptr<GG::Wnd> wnd) {
+    auto it = std::find(m_wnd_stack.begin(), m_wnd_stack.end(), wnd);
+    if (it != m_wnd_stack.end()) {
+        m_wnd_stack.erase(it);
     }
+}
 
-    if (m_production_wnd->Visible()) {
-        ToggleProduction();
-        some_subscreen_was_visible = true;
+bool MapWnd::ReturnToMap() {
+    std::shared_ptr<GG::Wnd> wnd;
+    while (!m_wnd_stack.empty() && !(wnd && wnd->Visible())) {
+        wnd = m_wnd_stack.back();
+        m_wnd_stack.pop_back();
     }
-
-    if (some_subscreen_was_visible)
-        return true;
-
-    if (m_combat_report_wnd->Visible()) {
-        m_combat_report_wnd->Hide();
+    if (!(wnd && wnd->Visible())) {
         return true;
     }
 
-    if (m_pedia_panel->Visible()) {
-        TogglePedia();
-        return true;
-    }
+    // prepare to close fleets window if open
+    auto& fm = FleetUIManager::GetFleetUIManager();
+    auto cui = ClientUI::GetClientUI();
 
-    if (m_object_list_wnd->Visible()) {
+    if (wnd == m_sitrep_panel) {
+        ToggleSitRep();
+    } else if (wnd == m_research_wnd) {
+        ToggleResearch();
+    } else if (wnd == m_design_wnd) {
+        ToggleDesign();
+    } else if (wnd == m_production_wnd) {
+        ToggleProduction();
+    } else if (wnd == m_pedia_panel) {
+        TogglePedia();
+    } else if (wnd == m_object_list_wnd) {
         ToggleObjects();
-        return true;
-    }
-
-    if (m_moderator_wnd->Visible()) {
+    } else if (wnd == m_moderator_wnd) {
         ToggleModeratorActions();
-        return true;
-    }
-
-    // close fleets window if open
-    FleetUIManager& fm = FleetUIManager::GetFleetUIManager();
-    GG::Wnd* active_fleet_wnd = fm.ActiveFleetWnd();
-    if (active_fleet_wnd) {
+    } else if (wnd == m_combat_report_wnd) {
+        m_combat_report_wnd->Hide();
+    } else if (wnd == m_side_panel) {
+        SelectSystem(INVALID_OBJECT_ID);
+    } else if (wnd.get() == fm.ActiveFleetWnd()) {
         fm.CloseAll();
-        return true;
-    }
-
-    // close sidepanel if open
-    if (SidePanel::SystemID() != INVALID_OBJECT_ID) {
-       SelectSystem(INVALID_OBJECT_ID);
-       return true;
+    } else if (cui && wnd == cui->GetPlayerListWnd()) {
+        HideEmpires();
+    } else if (cui && wnd == cui->GetMessageWnd()) {
+        HideMessages();
+    } else {
+        ErrorLogger() << "Unknown GG::Wnd " << wnd->Name() << " found in MapWnd::m_wnd_stack";
```

Comment:
are you sure `wnd` is not null here?

Reply:
Yes, the check on line 5865 will return from the function if `!wnd`